File: src/main/java/org/cytoscape/biogwplugin/internal/query/BGQueryFactory.java
package org.cytoscape.biogwplugin.internal.query;

import org.cytoscape.biogwplugin.internal.BGServiceManager;
import org.cytoscape.work.AbstractTask;
import org.cytoscape.work.AbstractTaskFactory;
import org.cytoscape.work.TaskIterator;

import java.util.ArrayList;

/**
 * Created by sholmas on 28/03/2017.
 */
public class BGQueryFactory extends AbstractTaskFactory {

    enum QueryType {
            NODE_FETCH, NODE_SEARCH, RELATION_SEARCH
    }

    private String serverUrl;
    private String queryString;
    private BGServiceManager serviceManager;
    private BGRelationsQuery.BGRelationDirection direction;
    private QueryType type;
    private ArrayList<Runnable> callbacks;

    public BGQueryFactory(String serverUrl, String queryString, BGServiceManager serviceManager, QueryType type) {
        this.serverUrl = serverUrl;
        this.queryString = queryString;
        this.serviceManager = serviceManager;
        this.type = type;
    }

    public BGQueryFactory(String serverUrl, String queryString, BGRelationsQuery.BGRelationDirection direction, BGServiceManager serviceManager, QueryType type) {
        this.serverUrl = serverUrl;
        this.queryString = queryString;
        this.direction = direction;
        this.type = type;
    }

    @Override
    public TaskIterator createTaskIterator() {
        BGQuery query = null;
        switch (this.type) {
            case NODE_FETCH:
                query = new BGFetchNodeQuery(serverUrl, queryString, serviceManager);
                break;
            case NODE_SEARCH:
                query = new BGNodeSearchQuery(serverUrl, queryString, serviceManager);
                break;
            case RELATION_SEARCH:
                // TODO: Warning, this will crash if the wrong constructor is used!
                assert direction != null;
                query = new BGRelationsQuery(serverUrl, queryString, direction);
                break;
            default:
                break;
        }
        query.setCallbacks(callbacks);
        TaskIterator taskIterator = new TaskIterator(query);
        return taskIterator;
    }

    public void setQueryString(String queryString) {
        this.queryString = queryString;
    }

    public void setType(QueryType type) {
        this.type = type;
    }

    public void setDirection(BGRelationsQuery.BGRelationDirection direction) {
        this.direction = direction;
    }

    public void addCallback(Runnable callback) {
        this.callbacks.add(callback);
    }
}

File: src/main/java/org/cytoscape/biogwplugin/internal/old/query/BGQueryFactory.kt
package org.cytoscape.biogwplugin.internal.old.query

import org.cytoscape.biogwplugin.internal.BGServiceManager
import org.cytoscape.work.AbstractTaskFactory
import org.cytoscape.work.TaskIterator

import java.util.ArrayList

/**
 * Created by sholmas on 28/03/2017.
 */
class BGQueryFactory(var serverUrl: String, var queryString: String, var serviceManager: BGServiceManager, var type: QueryType) : AbstractTaskFactory() {

    enum class QueryType {
        NODE_FETCH, NODE_SEARCH, RELATION_SEARCH, NOT_INITIALIZED
    }

    public constructor(serverUrl: String, queryString: String, serviceManager: BGServiceManager, type: QueryType, direction: BGRelationsQuery.Direction) : this(serverUrl, queryString, serviceManager, type) {
        this.direction = direction
    }

    var direction: BGRelationsQuery.Direction? = null

    private val callbacks = ArrayList<(BGQueryResult) -> Unit>()

    override fun createTaskIterator(): TaskIterator {
        val query: BGQuery = when (this.type) {
            BGQueryFactory.QueryType.NODE_FETCH ->  BGFetchNodeQuery(serverUrl, queryString, serviceManager)
            BGQueryFactory.QueryType.NODE_SEARCH -> BGNodeSearchQuery(serverUrl, queryString, serviceManager)
            BGQueryFactory.QueryType.RELATION_SEARCH -> {
                // TODO: Warning, this will crash if the wrong constructor is used!
                if(direction == null) throw Exception("Cannot run relation search without defining direction.")
                BGRelationsQuery(serverUrl, queryString, direction!!, serviceManager!!)
            }
            BGQueryFactory.QueryType.NOT_INITIALIZED -> throw Exception("Cannot run an un-initialized query.")
        }
        query.callbacks = callbacks
        val taskIterator = TaskIterator(query)
        return taskIterator
    }

    fun addCallback(callback: (BGQueryResult) -> Unit) {
        this.callbacks.add(callback)
    }
}
